File: lesson2/f.cpp
#include <algorithm>
#include <iostream>
#include <vector>

long long calculateWinnings(int n, const std::vector<long long>& sectors,
                            long long a, long long b, long long k) {
  long long max_winning = 0;
  long long max_sectors_right = b / k + (b % k != 0);
  long long max_finish_right = max_sectors_right % n;
  long long min_sectors_right = a / k + (a % k != 0);
  long long min_finish_right = min_sectors_right % n;
  if (max_sectors_right - min_sectors_right >= n) {
    max_winning = *std::max_element(sectors.begin(), sectors.end());
  } else {
    max_winning = *std::max_element(sectors.begin() + min_finish_right,
                                    sectors.begin() + max_finish_right + 1);
  }

  return max_winning;
}

int main() {
  long long n, a, b, k;
  std::cin >> n;
  std::vector<long long> sectors(n);
  for (int i = 0; i < n; i++) {
    std::cin >> sectors[i];
  }
  std::cin >> a >> b >> k;
  std::cout << calculateWinnings(n, sectors, a, b, k) << std::endl;
  return 0;
}

File: lesson2/f.py
from typing import List


def calculate_winnings(n: int, sectors: List[int], a: int, b: int, k: int) -> int:
    max_winning = 0
    max_sectors = b // k + (b % k != 0)  # 1   if b % k != 0 else b // k - 1
    max_finish_right = max_sectors % n - 1 if max_sectors % n != 0 else 0
    max_finish_left = n - max_finish_right
    min_sectors = a // k + (a % k != 0)  # 1   if a % k != 0 else a // k - 1
    min_finish_right = min_sectors % n - 1 if min_sectors % n != 0 else 0
    min_finish_left = n - min_finish_right if min_finish_right != 0 else 0
    if max_sectors - min_sectors >= n - 1:
        max_winning = max(sectors)
    else:
        for i in range(max_sectors-min_sectors+1):
            index = (i + min_finish_right) % n
            max_winning = max(max_winning, sectors[index])
            index = min_finish_left-i
            max_winning = max(max_winning, sectors[index])

    return max_winning


if __name__ == "__main__":
    n = int(input())
    sectors = list(map(int, input().split()))
    a, b, k = map(int, input().split())
    print(calculate_winnings(n, sectors, a, b, k))
